src/crimson/os/seastore/journal/segment_allocator.cc

```diff
@@ -250,17 +249,24 @@ SegmentAllocator::close_segment()
   bl.append(bp);
 
   assert(bl.length() == sm_group.get_rounded_tail_length());
-  return seg_to_close->write(
-    sm_group.get_segment_size() - sm_group.get_rounded_tail_length(),
-    bl
-  ).safe_then([seg_to_close=std::move(seg_to_close)] {
-    return seg_to_close->close();
+  return seg_to_close->advance_wp(
+    sm_group.get_segment_size() - sm_group.get_rounded_tail_length()
+  ).safe_then([this, FNAME, bl, seg_to_close=std::move(seg_to_close)] {
+    DEBUG("Writing tail info to segment {}", seg_to_close->get_segment_id());
+    return seg_to_close->write(
+      sm_group.get_segment_size() - sm_group.get_rounded_tail_length(),
+      bl
+    ).safe_then([this, seg_to_close=std::move(seg_to_close)] {
+      return seg_to_close->close(
+      ).safe_then([this, seg_to_close=std::move(seg_to_close)] {
+        segment_provider.close_segment(seg_to_close->get_segment_id());
+      });
+    });
```

Comment:
Some issues to note here:
* moving `seg_to_close` into nested labmda needs to mark the capture `mutable`, as moving needs to mutate the original captured variable.
* move `bl` wherever possible to reduce copies.
* flat lambdas are more likely to be also simpler and cleaner.

For example,
```c++
auto p_seg_to_close = seg_to_close.get();
return p_seg_to_close->advance_wp(
  sm_group.get_segment_size() - sm_group.get_rounded_tail_length()
).safe_then([this, FNAME, bl=std::move(bl), p_seg_to_close]() mutable {
  DEBUG("Writing tail info to segment {}", p_seg_to_close->get_segment_id());
  return p_seg_to_close->write(
    sm_group.get_segment_size() - sm_group.get_rounded_tail_length(),
    std::move(bl));
}).safe_then([p_seg_to_close] {
  return p_seg_to_close->close();
}).safe_then([this, seg_to_close=std::move(seg_to_close)] {
  segment_provider.close_segment(seg_to_close->get_segment_id());
}) ...
```

Reply:
Ok, Done.